tools/transforms_fuzzer.cc

```diff
@@ -0,0 +1,113 @@
+    pixel_type* row = p->Row(y);
+    for (size_t x = 0; x < w; ++x) {
+      row[x] = rng() & 0x7FFFFFFF;
+    }
+  }
+}
+}  // namespace
+
+int TestOneInput(const uint8_t* data, size_t size) {
+  static Status nevermind = true;
+  BitReader reader(Span<const uint8_t>(data, size));
+  BitReaderScopedCloser reader_closer(&reader, &nevermind);
+
+  std::mt19937 rng(reader.ReadFixedBits<56>());
+
+  // One of {0, 1, _2_, 3}; "2" will be filtered out soon.
+  size_t nb_chans = static_cast<size_t>(reader.ReadFixedBits<8>()) & 0x3;
+  size_t nb_extra = static_cast<size_t>(reader.ReadFixedBits<8>()) & 0x7;
+  // 1..32
+  size_t bit_depth =
+      (static_cast<size_t>(reader.ReadFixedBits<8>()) & 0x1F) + 1;
+  // {0, 1, 2, 3}
+  size_t log_upsampling =
+      (static_cast<size_t>(reader.ReadFixedBits<8>()) & 0x3);
+  size_t upsampling = 1 << log_upsampling;
+
+  size_t w_orig = static_cast<size_t>(reader.ReadFixedBits<16>());
+  size_t h_orig = static_cast<size_t>(reader.ReadFixedBits<16>());
+  size_t w = DivCeil(w_orig, upsampling);
+  size_t h = DivCeil(h_orig, upsampling);
+
+  if ((nb_chans == 2) || ((nb_chans + nb_extra) == 0) || (w * h == 0) ||
+      ((w_orig * h_orig * (nb_chans + nb_extra)) > (1 << 23))) {
+    return 0;
+  }
+
+  Image image(w, h, bit_depth, nb_chans + nb_extra);
+
+  for (size_t c = 0; c < nb_chans; c++) {
+    image.channel[c].hshift =
+        static_cast<size_t>(reader.ReadFixedBits<8>()) & 1;
+    image.channel[c].vshift =
+        static_cast<size_t>(reader.ReadFixedBits<8>()) & 1;
+    size_t xsize_shifted = DivCeil(w, 1 << image.channel[c].hshift);
+    size_t ysize_shifted = DivCeil(h, 1 << image.channel[c].vshift);
+    image.channel[c].shrink(xsize_shifted, ysize_shifted);
+  }
+
+  for (size_t ec = 0, c = nb_chans; ec < nb_extra; ec++, c++) {
+    size_t log_ec_upsampling =
+        (static_cast<size_t>(reader.ReadFixedBits<8>()) & 0x3);
+    log_ec_upsampling = std::max(log_ec_upsampling, log_upsampling);
+    size_t ec_upsampling = 1 << log_ec_upsampling;
+    image.channel[c].shrink(DivCeil(w_orig, ec_upsampling),
+                            DivCeil(h_orig, ec_upsampling));
+    image.channel[c].hshift = image.channel[c].vshift =
+        CeilLog2Nonzero(ec_upsampling) - CeilLog2Nonzero(upsampling);
+  }
+
+  GroupHeader header;
+  if (!Bundle::Read(&reader, &header)) return 0;
+  weighted::Header w_header;
+  if (!Bundle::Read(&reader, &w_header)) return 0;
+
+  // TODO(eustas): give it a try?
+  if (!reader.AllReadsWithinBounds()) return 0;
+
+  image.transform = header.transforms;
+  for (Transform& transform : image.transform) {
+    if (!transform.MetaApply(image)) return 0;
+  }
+  if (image.error) return 0;
+
+  ModularOptions options;
+  if (!ValidateChannelDimensions(image, options)) return 0;
+
+  for (size_t i = 0; i < image.channel.size(); ++i) {
+    FillChannel(image.channel[i], rng);
+  }
+
+  image.undo_transforms(w_header, /*keep=*/-1);
+
+  if (image.error) return 0;
+
```

Comment:
We should also check that at this point the channels match up with the original channels - so no meta channels, and all sizes are the same.

Reply:
Sure. Going to check that image.error should also be false here (all in all, we did no crime...)